library/src/scripts/flyouts/NewPostMenu.tsx

```diff
@@ -59,47 +71,285 @@ function ActionItem({ item, style }: { item: IAddPost; style: ITransition }) {
+    const vars = newPostMenuVariables();
+
+    let { items } = props;
+
+    const buttonRef = useRef<HTMLButtonElement>(null);
+    const backgroundRef = useRef<HTMLElement>(null);
+    const menuRef = useRef<HTMLUListElement>(null);
+    const itemsRef = useRef<HTMLLIElement>(null);
+    const accessMenuRef = useRef<HTMLUListElement>(null);
+
+    const [state, dispatch] = useReducer(reducer, initialState);
+
+    const onClickBackground = (event: React.MouseEvent) => {
+        event.stopPropagation();
+        if (state.open) {
+            dispatch({ type: "set_button_focus", focus: true });
+            dispatch({ type: "set_open", open: false });
+            dispatch({ type: "set_focused_item", item: undefined });
         }
     };
 
-    const classes = newPostMenuClasses();
-    const { items } = props;
+    const onBgKeyDown = (event: React.KeyboardEvent<any>) => {
+        event.preventDefault();
+        event.stopPropagation();
+
+        switch (event.key) {
+            case "Escape":
+                if (state.open) {
+                    dispatch({ type: "set_open", open: false });
+                    dispatch({ type: "set_button_focus", focus: true });
+                    dispatch({ type: "set_focused_item", item: undefined });
+                }
+                break;
+            default:
+                // We don't want to lose focus in case we already have
+                if (state.buttonFocus) {
+                    dispatch({ type: "set_button_focus", focus: true });
+                }
+                break;
+        }
+    };
+
+    const onKeyDown = (event: React.KeyboardEvent<any>) => {
+        event.preventDefault();
+        event.stopPropagation();
+
+        switch (event.key) {
+            case "Escape":
+                if (state.open) {
+                    dispatch({ type: "set_open", open: false });
+                    dispatch({ type: "set_button_focus", focus: true });
+                    dispatch({ type: "set_focused_item", item: undefined });
+                }
+                break;
+            case "Home":
+                if (state.open) {
+                    dispatch({ type: "set_focused_item", item: 0 });
+                }
+                break;
+            case "End":
+                if (state.open) {
+                    dispatch({ type: "set_focused_item", item: items.length - 1 });
+                }
+                break;
+            case "ArrowUp":
+                if (state.open && typeof state.focusedItem != "undefined") {
+                    dispatch({ type: "set_focused_item", item: (state.focusedItem + 1) % items.length });
+                }
+                break;
+            case "ArrowDown":
+                if (state.open && typeof state.focusedItem != "undefined") {
+                    dispatch({ type: "set_focused_item", item: (state.focusedItem - 1 + items.length) % items.length });
+                }
+                break;
```

Comment:
You've got the up and down arrows mixed up here :)

Reply:
I think it's consistent with the way the items are counted numerically (but visually it might be confusing yes).